lib/string_decoder.js

```diff
@@ -2,224 +2,234 @@
-      this.surrogateSize = 3;
-      this.detectIncompleteChar = base64DetectIncompleteChar;
+      this.text = base64Text;
+      this.end = base64End;
+      nb = 3;
       break;
     default:
-      this.write = passThroughWrite;
+      this.write = simpleWrite;
+      this.end = simpleEnd;
       return;
   }
+  this.lastNeed = 0;
+  this.lastTotal = 0;
+  this.lastChar = Buffer.allocUnsafe(nb);
+}
 
-  // Enough space to store all bytes of a single character. UTF-8 needs 4
-  // bytes, but CESU-8 may require up to 6 (3 bytes per surrogate).
-  this.charBuffer = Buffer.allocUnsafe(6);
-  // Number of bytes received for the current incomplete multi-byte character.
-  this.charReceived = 0;
-  // Number of bytes expected for the current incomplete multi-byte character.
-  this.charLength = 0;
-};
-
-
-// write decodes the given buffer and returns it as JS string that is
-// guaranteed to not contain any partial multi-byte characters. Any partial
-// character found at the end of the buffer is buffered up, and will be
-// returned when calling write again with the remaining bytes.
-//
-// Note: Converting a Buffer containing an orphan surrogate to a String
-// currently works, but converting a String to a Buffer (via `Buffer.from()`,
-// or Buffer#write) will replace incomplete surrogates with the unicode
-// replacement character. See https://codereview.chromium.org/121173009/ .
-StringDecoder.prototype.write = function(buffer) {
-  var charStr = '';
-  var buflen = buffer.length;
-  var charBuffer = this.charBuffer;
-  var charLength = this.charLength;
-  var charReceived = this.charReceived;
-  var surrogateSize = this.surrogateSize;
-  var encoding = this.encoding;
-  var charCode;
-  // if our last write ended with an incomplete multibyte character
-  while (charLength) {
-    // determine how many remaining bytes this buffer has to offer for this char
-    var diff = charLength - charReceived;
-    var available = (buflen >= diff) ? diff : buflen;
-
-    // add the new bytes to the char buffer
-    buffer.copy(charBuffer, charReceived, 0, available);
-    charReceived += available;
-
-    if (charReceived < charLength) {
-      // still not enough chars in this buffer? wait for more ...
-
-      this.charLength = charLength;
-      this.charReceived = charReceived;
-
+StringDecoder.prototype.write = function(buf) {
+  if (buf.length === 0)
+    return '';
+  var r;
+  var i;
+  if (this.lastNeed) {
+    r = this.fillLast(buf);
+    if (r === undefined)
       return '';
-    }
-
-    // remove bytes belonging to the current character from the buffer
-    buffer = buffer.slice(available, buflen);
-    buflen = buffer.length;
-
-    // get the character that was split
-    charStr = charBuffer.toString(encoding, 0, charLength);
+    i = this.lastNeed;
+    this.lastNeed = 0;
+  } else {
+    i = 0;
+  }
+  if (i < buf.length)
+    return (r ? r + this.text(buf, i) : this.text(buf, i));
```

Comment:
Is the ternary necessary?  `r` is always an empty string here, if I read it right.